torchdata/datapipes/iter/transform/callable.py

```diff
@@ -205,3 +206,162 @@ def __len__(self) -> int:
+                    new_item = {k: v for (k, v) in old_item.items() if k in new_keys}  # type: ignore[assignment]
+            else:
+                new_item = old_item
+                warnings.warn(
+                    "The next item was not an iterable and cannot be filtered, "
+                    "please be aware that no filter was done or new item created."
+                )
+
+            if isinstance(self.index, list):
+                # check to make sure all indices requested were in the item. warn if not
+                try:
+                    for i in self.index:
+                        old_item[i]
+                except (IndexError, KeyError):
+                    warnings.warn(
+                        "At least one index in the filter is not present in the item being returned,"
+                        " please be aware that expected columns/keys may be missing."
+                    )
+
+            yield new_item  # type: ignore[misc]
+
+    def __len__(self) -> int:
+        if isinstance(self.datapipe, Sized):
+            return len(self.datapipe)
+        raise TypeError(f"{type(self).__name__} instance doesn't have valid length")
+
+
+@functional_datapipe("flatten")
+class FlattenIterDataPipe(IterDataPipe[T_co]):
+    r"""
+    returns a flattened copy of the input DataPipe based on provided indices (functional name: ``flatten``).
+
+    Args:
+        datapipe: IterDataPipe with iterable elements
+        indices: a single index/key for the item to flatten from an iterator item or a list of indices/keys to be flattened
+
+    Example:
+        >>> from torchdata.datapipes.iter import IterableWrapper
+        >>> dp = IterableWrapper([(0, 10, (100, 1000)), (1, 11, (111, 1001)), (2, 12, (122, 1002)), (3, 13, (133, 1003)), (4, 14, (144, 1004))])
+        >>> flatten_dp = dp.flatten(2)
+        >>> list(flatten_dp)
+        [(0, 10, 100, 1000), (1, 11, 111, 1001), (2, 12, 122, 1002), (3, 13, 133, 1003), (4, 14, 144, 1004)]
+    """
+    datapipe: IterDataPipe
+    indices = None
+
+    def __init__(
+        self,
+        datapipe: IterDataPipe,
+        indices: Optional[Union[Hashable, List[Hashable]]] = None,
+    ) -> None:
+        super().__init__()
+        self.datapipe = datapipe
+        if indices:
+            if isinstance(indices, list):
+                self.indices = set(indices)
+            else:
+                self.indices = {indices}
+
+    def __iter__(self) -> Iterator[T_co]:
+        for old_item in self.datapipe:
+            if self.indices:
+                if isinstance(old_item, dict):
+                    new_item = {}  # type: ignore[assignment]
+                    for k, v in old_item.items():
+                        if k in self.indices:
+                            for k_sub, v_sub in v.items():
+                                new_item[k_sub] = v_sub
+                        else:
+                            new_item[k] = v
```

Comment:
Wdyt about adding a warning when `k_sub` has already been presented in `old_item`?

Reply:
yep, good idea. i can do that